src/main/java/org/apache/cassandra/sidecar/restore/RestoreSliceTask.java

```diff
@@ -430,22 +450,79 @@ private Future<File> validateFiles(File directory)
             }
 
             // all files match with the provided checksums
-            promise.tryComplete(directory);
+            return directory;
         }, false); // unordered
 
         return StopWatch.measureTimeTaken(future, d -> instanceMetrics.restore().sliceValidationTime.metric.update(d, TimeUnit.NANOSECONDS));
     }
 
-    private void compareChecksums(Map<String, String> expectedChecksums, File[] files, Promise<?> promise)
+    // Remove all the SSTables that does not belong this node
+    // The method modifies the input manifest and delete files under directory, if out of range sstables are found
+    private void removeOutOfRangeSSTables(File directory, RestoreSliceManifest manifest) throws RestoreJobException, IOException
+    {
+        Set<TokenRange> ranges = localTokenRangesProvider.localTokenRanges(slice.keyspace()).get(slice.owner().id());
+        if (ranges == null || ranges.isEmpty())
+        {
+            // Note: retry is allowed for the failure
+            throw new RestoreJobException("Unable to fetch local range, retry later");
+        }
+
+        // 1. remove the sstables that are fully out of range
+        // 2. detect if there is any range that partially overlaps. In that case, signal that this node is required to run nodetool cleanup on job completion
+        Iterator<Map.Entry<String, RestoreSliceManifest.ManifestEntry>> it = manifest.entrySet().iterator();
+        while (it.hasNext())
+        {
+            RestoreSliceManifest.ManifestEntry entry = it.next().getValue();
+            // TokenRange is open-closed, hence subtracting one from the rangeStart read from manifest
+            TokenRange sstableRange = new TokenRange(entry.startToken().subtract(BigInteger.ONE),
+                                                     entry.endToken());
+
+            boolean hasOverlap = false;
+            boolean fullyEnclosed = false;
+            for (TokenRange owningRange : ranges)
+            {
+                if (hasOverlap)
+                {
+                    break;
+                }
+
+                hasOverlap = owningRange.overlaps(sstableRange);
+
+                if (hasOverlap)
+                {
+                    fullyEnclosed = owningRange.encloses(sstableRange);
+                }
+            }
+
+            // fully out of range
+            if (!hasOverlap)
+            {
+                // remove the entry from manifest
+                it.remove();
+                // delete the files
+                for (String fileName : entry.componentsChecksum().keySet())
+                {
+                    Path path = directory.toPath().resolve(fileName);
+                    Files.deleteIfExists(path);
+                }
```

Comment:
Nit: Maybe extracting this to a private method? It may help with readability (the method is already big)